File: swift/LoudKit/Safetensors.swift
import Foundation

/// Minimal safetensors reader: 8-byte little-endian header length, a JSON
/// header mapping tensor names to `{dtype, shape, data_offsets}`, then the
/// raw payload. The file is memory-mapped, so opening a 1.3 GB checkpoint
/// costs nothing until tensors are actually read, the same lazy contract as
/// `loudkit.checkpoint.Checkpoint`.
public final class Safetensors {
    /// Where one tensor is in the payload, and what shape it claims.
    public struct TensorInfo: Sendable, Equatable {
        /// The safetensors dtype string, `F32` and `I64` among them.
        public let dtype: String
        /// Dimensions, outermost first.
        public let shape: [Int]
        let begin: Int
        let end: Int
    }

    private let data: Data
    private let payloadOffset: Int
    /// Every tensor in the file, by name.
    public let tensors: [String: TensorInfo]
    /// The header's `__metadata__` block. The checkpoint manifest lives here.
    public let metadata: [String: String]

    /// Map a safetensors file and parse its header, reading no tensor.
    ///
    /// - Throws: `LoudKitError.asset` on a truncated file, a header that is not
    ///   JSON, or an offset that would read outside the payload.
    public init(url: URL) throws {
        data = try Data(contentsOf: url, options: .mappedIfSafe)
        guard data.count > 8 else { throw LoudKitError.asset("\(url.lastPathComponent): truncated") }
        let headerLen = data.subdata(in: 0..<8).withUnsafeBytes {
            $0.loadUnaligned(as: UInt64.self).littleEndian
        }
        // Bounds-checked as UInt64 *before* the conversion. `Int(headerLen)`
        // traps for anything >= 2^63, so a corrupt or hostile file did not get
        // the refusal below, it killed the process with SIGTRAP, inside the
        // check meant to protect against it. This reader is reached by
        // `VoiceProfile.load`, i.e. by any downloaded `.voice.safetensors`.
        // Rust, Go and JS all survive the same header. The existing test used
        // `UInt64(1 << 40)`, which is under `Int.max`, so it passed.
        guard headerLen <= UInt64(data.count - 8) else {
            throw LoudKitError.asset("\(url.lastPathComponent): header overruns file")
        }
        payloadOffset = 8 + Int(headerLen)
        let headerData = data.subdata(in: 8..<payloadOffset)
        // Last-wins on a repeated name, like the reference. See
        // `ManifestReader.json`: a repeated tensor entry here picks a
        // different slice of the payload in this port than in the other four.
        let headerRaw = try ManifestReader.json(
            headerData, "\(url.lastPathComponent): bad header JSON")
        guard let header = headerRaw as? [String: Any] else {
            throw LoudKitError.asset(
                "\(url.lastPathComponent): header is "
                + "\(ManifestReader.jsonTypePhrase(headerRaw)), expected a JSON object")
        }
        var readTensors: [String: TensorInfo] = [:]
        var readMetadata: [String: String] = [:]
        for (name, value) in header {
            if name == "__metadata__" {
                readMetadata = value as? [String: String] ?? [:]
                continue
            }
            guard let entry = value as? [String: Any],
                  let dtype = entry["dtype"] as? String,
                  let shape = (entry["shape"] as? [NSNumber])?.map({ $0.intValue }),
                  let offsets = (entry["data_offsets"] as? [NSNumber])?.map({ $0.intValue }),
                  offsets.count == 2 else {
                throw LoudKitError.asset("\(url.lastPathComponent): bad entry for \(name)")
            }
            // Validated here, where the other header checks live, because
            // `raw` slices `data[begin..<end]` and a Data subscript out of
            // range traps, the process dies rather than throwing. The offsets
            // come straight from the file, so a truncated or corrupt
            // checkpoint must be an error, not a crash.
            // Bounded *before* the addition, for the same reason `headerLen`
            // is bounded before its conversion a few lines up: Swift traps on
            // integer overflow, so `data_offsets: [9223372036854775807, …]`
            // killed the process inside the check written to refuse it. The
            // remaining budget is `data.count - payloadOffset`, and
            // `payloadOffset <= data.count` is already established, so this
            // subtraction cannot go negative and neither addition can overflow.
            guard offsets[0] >= 0, offsets[1] >= offsets[0],
                  offsets[1] <= data.count - payloadOffset else {
                throw LoudKitError.asset(
                    "\(url.lastPathComponent): \(name) has offsets outside the file")
            }
            let begin = offsets[0] + payloadOffset
            let end = offsets[1] + payloadOffset
            guard begin <= end, end <= data.count else {
                throw LoudKitError.asset(
                    "\(url.lastPathComponent): tensor \(name) spans \(offsets[0])..<\(offsets[1]) "
                        + "of a \(data.count - payloadOffset)-byte payload: the file is truncated "
                        + "or the header is corrupt")
            }
            // The shape must account for exactly the bytes claimed.
            //
            // The range check above keeps `raw` from trapping, but callers
            // read `shape` to size their work: `VoiceProfile` checks a
            // dimension and `TokenGenerator` passes m/n/k straight to
            // `cblas_sgemm`, which reads what it is told to read. A header
            // declaring shape [256] over four bytes of payload is therefore
            // not a bad tensor, it is an out-of-bounds read in C with no
            // bounds check anywhere on the path. Overflow is checked rather
            // than assumed: `[1<<40, 1<<40]` multiplies to something small and
            // plausible if it is allowed to wrap.
            var elements = 1
            for dim in shape {
                guard dim >= 0 else {
                    throw LoudKitError.asset(
                        "\(url.lastPathComponent): tensor \(name) has a negative dimension in "
                            + "\(shape)")
                }
                let (product, overflowed) = elements.multipliedReportingOverflow(by: dim)
                guard !overflowed else {
                    throw LoudKitError.asset(
                        "\(url.lastPathComponent): tensor \(name) shape \(shape) overflows")
                }
                elements = product
            }
            guard let width = Self.byteWidth(of: dtype) else {
                throw LoudKitError.asset(
                    "\(url.lastPathComponent): tensor \(name) has unknown dtype \(dtype)")
            }
            let (declared, widthOverflowed) = elements.multipliedReportingOverflow(by: width)
            guard !widthOverflowed, declared == end - begin else {
                throw LoudKitError.asset(
                    "\(url.lastPathComponent): tensor \(name) declares shape \(shape) of "
                        + "\(dtype) (\(widthOverflowed ? "overflow" : String(declared)) bytes) but "
                        + "occupies \(end - begin) bytes: the header does not describe the payload")
            }
            readTensors[name] = TensorInfo(dtype: dtype, shape: shape, begin: begin, end: end)
        }
        tensors = readTensors
        metadata = readMetadata
    }

    /// Bytes per element, or nil for a dtype this reader does not know.
    ///
    /// Listed rather than inferred: an unknown dtype must be refused at load,
    /// not discovered later by a reader that happens to be asked for it.
    static func byteWidth(of dtype: String) -> Int? {
        switch dtype {
        case "F64", "I64", "U64": return 8
        case "F32", "I32", "U32": return 4
        case "F16", "BF16", "I16", "U16": return 2
        case "I8", "U8", "BOOL": return 1
        default: return nil
        }
    }

    /// Tensor names beginning with `prefix`, sorted. An empty prefix is all
    /// of them.
    public func names(prefix: String = "") -> [String] {
        tensors.keys.filter { $0.hasPrefix(prefix) }.sorted()
    }

    private func raw(_ name: String) throws -> (TensorInfo, Data) {
        guard let info = tensors[name] else {
            throw LoudKitError.asset("tensor not found: \(name)")
        }
        return (info, data.subdata(in: info.begin..<info.end))
    }

    /// Tensor as `[Float]`, upcasting F16 storage exactly (fp16 -> fp32 is a
    /// widening conversion, so the packed value survives untouched).
    public func floats(_ name: String) throws -> [Float] {
        let (info, bytes) = try raw(name)
        switch info.dtype {
        // `loadUnaligned`, not `bindMemory`. A tensor starts wherever the
        // header ends, so these slices carry no alignment guarantee, and
        // `bindMemory` requires one. It works on ARM, which tolerates unaligned
        // loads, and is undefined behaviour by Swift's own rules: the optimiser
        // is entitled to assume the alignment it was promised. A safetensors
        // file is data from outside the process; this is the one reader that
        // sees it first.
        case "F32":
            return bytes.withUnsafeBytes { buf in
                (0..<(buf.count / MemoryLayout<Float32>.size)).map {
                    buf.loadUnaligned(
                        fromByteOffset: $0 * MemoryLayout<Float32>.size, as: Float32.self)
                }
            }
        case "F16":
            return bytes.withUnsafeBytes { buf in
                (0..<(buf.count / MemoryLayout<Float16>.size)).map {
                    Float(
                        buf.loadUnaligned(
                            fromByteOffset: $0 * MemoryLayout<Float16>.size, as: Float16.self))
                }
            }
        default:
            throw LoudKitError.asset("\(name): dtype \(info.dtype) is not a float type this reader converts")
        }
    }

    /// One integer tensor, flattened, whatever width it was stored at.
    ///
    /// - Throws: `LoudKitError.asset` when the name is absent or the dtype is
    ///   not an integer one.
    public func ints(_ name: String) throws -> [Int] {
        let (info, bytes) = try raw(name)
        switch info.dtype {
        // `loadUnaligned` here for the same reason as `floats` above, and this
        // reader was left on `bindMemory` when that one was fixed: a tensor
        // starts wherever the header ends, so the slice carries no alignment
        // guarantee and `bindMemory` requires one. It happens to work on ARM
        // and is undefined behaviour by Swift's own rules, the optimiser is
        // entitled to assume the alignment it was promised, and a safetensors
        // file is data from outside the process.
        case "I64":
            return bytes.withUnsafeBytes { buf in
                (0..<(buf.count / MemoryLayout<Int64>.size)).map {
                    Int(
                        buf.loadUnaligned(
                            fromByteOffset: $0 * MemoryLayout<Int64>.size, as: Int64.self))
                }
            }
        case "I32":
            return bytes.withUnsafeBytes { buf in
                (0..<(buf.count / MemoryLayout<Int32>.size)).map {
                    Int(
                        buf.loadUnaligned(
                            fromByteOffset: $0 * MemoryLayout<Int32>.size, as: Int32.self))
                }
            }
        default:
            throw LoudKitError.asset("\(name): dtype \(info.dtype) is not an int type this reader converts")
        }
    }

    /// One tensor's declared shape, without reading its bytes.
    ///
    /// - Throws: `LoudKitError.asset` when the name is absent.
    public func shape(_ name: String) throws -> [Int] {
        guard let info = tensors[name] else {
            throw LoudKitError.asset("tensor not found: \(name)")
        }
        return info.shape
    }

    /// One tensor to write: its name, dtype, shape and little-endian bytes.
    public struct Entry: Sendable {
        /// The tensor's name in the file.
        public let name: String
        /// The safetensors dtype string to write.
        public let dtype: String
        /// Dimensions, outermost first. Their product must match `data`.
        public let shape: [Int]
        /// The payload, little-endian, already in the dtype's width.
        public let data: Data

        /// The four fields are all required: a writer that defaulted any of
        /// them could write a header that does not describe its own bytes.
        public init(name: String, dtype: String, shape: [Int], data: Data) {
            self.name = name
            self.dtype = dtype
            self.shape = shape
            self.data = data
        }
    }

    /// The order the safetensors library lays tensors out in: widest dtype
    /// first, then by name. Only the order the reader inverts.
    private static let dtypeRank = [
        "BOOL", "U8", "I8", "F8_E5M2", "F8_E4M3", "I16", "U16", "F16", "BF16",
        "I32", "U32", "F32", "F64", "I64", "U64"
    ]

    /// Lay out tensors and metadata as a safetensors file: the 8-byte header
    /// length, the JSON header padded to a multiple of eight with spaces, then
    /// the payloads in header order. Owner-only permissions.
    public static func write(_ entries: [Entry], metadata: [String: String], to url: URL) throws {
        let sorted = entries.sorted { a, b in
            let (ra, rb) = (dtypeRank.firstIndex(of: a.dtype) ?? 0, dtypeRank.firstIndex(of: b.dtype) ?? 0)
            return ra != rb ? ra > rb : a.name < b.name
        }
        var fields: [String] = []
        if !metadata.isEmpty {
            let meta = try JSONSerialization.data(withJSONObject: metadata, options: [.sortedKeys])
            guard let text = String(bytes: meta, encoding: .utf8) else {
                throw LoudKitError.asset("the safetensors metadata is not UTF-8")
            }
            fields.append("\"__metadata__\":" + text)
        }
        var offset = 0
        for e in sorted {
            guard let width = byteWidth(of: e.dtype) else {
                throw LoudKitError.asset("\(e.name): unknown dtype \(e.dtype)")
            }
            let elements = e.shape.reduce(1, *)
            guard elements * width == e.data.count else {
                throw LoudKitError.shape(
                    "\(e.name): shape \(e.shape) of \(e.dtype) is \(elements * width) bytes, "
                    + "data is \(e.data.count)")
            }
            guard let quoted = String(
                bytes: try JSONSerialization.data(withJSONObject: [e.name]), encoding: .utf8) else {
                throw LoudKitError.asset("\(e.name): the tensor name is not UTF-8")
            }
            let shape = e.shape.map(String.init).joined(separator: ",")
            fields.append(
                "\(quoted.dropFirst().dropLast()):{\"dtype\":\"\(e.dtype)\",\"shape\":[\(shape)],"
                + "\"data_offsets\":[\(offset),\(offset + e.data.count)]}")
            offset += e.data.count
        }
        var header = "{" + fields.joined(separator: ",") + "}"
        while header.utf8.count % 8 != 0 { header += " " }
        var out = Data()
        var length = UInt64(header.utf8.count).littleEndian
        withUnsafeBytes(of: &length) { out.append(contentsOf: $0) }
        out.append(contentsOf: Array(header.utf8))
        for e in sorted { out.append(e.data) }
        try out.write(to: url)
        try? FileManager.default.setAttributes([.posixPermissions: 0o600], ofItemAtPath: url.path)
    }
}

File: go/safetensors/safetensors.go
// Package safetensors reads the loudkit checkpoint and voice files — enough
// of the format to pull the embedding tables and the manifest. 8-byte
// little-endian header length, a JSON header naming each tensor with its
// dtype, shape and byte offsets, then the raw tensors.
package safetensors

import (
	"encoding/binary"
	"encoding/json"
	"fmt"
	"math"
	"os"
)

// Tensor is one named tensor: its shape and a byte slice of its data.
type Tensor struct {
	Dtype string
	Shape []int64
	Data  []byte
}

// File is a parsed safetensors container.
type File struct {
	Tensors  map[string]Tensor
	Metadata map[string]string
}

// Open reads and parses a safetensors file.
func Open(path string) (*File, error) {
	buf, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	if len(buf) < 8 {
		return nil, fmt.Errorf("%s: too small to be safetensors", path)
	}
	headerLen := binary.LittleEndian.Uint64(buf[:8])
	if headerLen > uint64(len(buf)-8) {
		return nil, fmt.Errorf("%s: bad header length %d", path, headerLen)
	}
	var header map[string]interface{}
	if err := json.Unmarshal(buf[8:8+headerLen], &header); err != nil {
		return nil, fmt.Errorf("%s: bad header JSON: %w", path, err)
	}
	f := &File{
		Tensors:  map[string]Tensor{},
		Metadata: map[string]string{},
	}
	base := 8 + headerLen
	for name, raw := range header {
		if name == "__metadata__" {
			meta, ok := raw.(map[string]interface{})
			if !ok {
				continue
			}
			for k, v := range meta {
				if s, ok := v.(string); ok {
					f.Metadata[k] = s
				}
			}
			continue
		}
		spec, ok := raw.(map[string]interface{})
		if !ok {
			continue
		}
		dtype, _ := spec["dtype"].(string)
		offsets, _ := spec["data_offsets"].([]interface{})
		if len(offsets) != 2 {
			return nil, fmt.Errorf("%s: tensor %q missing data_offsets", path, name)
		}
		begin := uint64(toFloat(offsets[0]))
		end := uint64(toFloat(offsets[1]))
		shape := []int64{}
		if shapeRaw, ok := spec["shape"].([]interface{}); ok {
			for _, d := range shapeRaw {
				shape = append(shape, int64(toFloat(d)))
			}
		}
		// Validate against base+end, not end: the slice below is indexed from
		// the end of the header, so a tensor whose offsets fit inside the
		// file but not inside the *data section* would slice out of range and
		// panic. base+end can also wrap on a hostile header, hence the
		// overflow check.
		if begin > end || base+end < base || base+end > uint64(len(buf)) {
			return nil, fmt.Errorf("%s: tensor %q out of range", path, name)
		}
		// The shape must account for exactly the bytes claimed.
		//
		// The range check above stops a slice panic, but callers read Shape to
		// size their work — a header declaring [256] over four bytes of payload
		// is not a bad tensor, it is a reader that computes with a length the
		// data does not have. The typed accessors below also divided by the
		// element width and silently dropped a partial tail; with this check a
		// partial tail cannot exist.
		width, ok := byteWidth(dtype)
		if !ok {
			return nil, fmt.Errorf("%s: tensor %q has unknown dtype %q", path, name, dtype)
		}
		elements := int64(1)
		for _, dim := range shape {
			if dim < 0 {
				return nil, fmt.Errorf("%s: tensor %q has a negative dimension in %v",
					path, name, shape)
			}
			if dim != 0 && elements > math.MaxInt64/dim {
				return nil, fmt.Errorf("%s: tensor %q shape %v overflows", path, name, shape)
			}
			elements *= dim
		}
		if declared := elements * int64(width); declared != int64(end-begin) {
			return nil, fmt.Errorf(
				"%s: tensor %q declares shape %v of %s (%d bytes) but occupies %d bytes — "+
					"the header does not describe the payload",
				path, name, shape, dtype, declared, end-begin)
		}
		f.Tensors[name] = Tensor{
			Dtype: dtype,
			Shape: shape,
			Data:  buf[base+begin : base+end],
		}
	}
	return f, nil
}

// byteWidth is bytes per element, and false for a dtype this reader does not
// know. Listed rather than inferred: an unknown dtype must be refused at load,
// not discovered later by whichever accessor is asked for it first.
func byteWidth(dtype string) (int, bool) {
	switch dtype {
	case "F64", "I64", "U64":
		return 8, true
	case "F32", "I32", "U32":
		return 4, true
	case "F16", "BF16", "I16", "U16":
		return 2, true
	case "I8", "U8", "BOOL":
		return 1, true
	}
	return 0, false
}

func toFloat(x interface{}) float64 {
	switch v := x.(type) {
	case float64:
		return v
	case int:
		return float64(v)
	}
	return 0
}

// F32 returns a tensor's data as []float32 (F32, or F16 upcast exactly).
func (f *File) F32(name string) ([]float32, error) {
	t, ok := f.Tensors[name]
	if !ok {
		return nil, fmt.Errorf("no tensor %q", name)
	}
	switch t.Dtype {
	case "F32":
		if len(t.Data)%4 != 0 {
			return nil, fmt.Errorf("%s: F32 data not 4-aligned", name)
		}
		out := make([]float32, len(t.Data)/4)
		for i := range out {
			out[i] = math.Float32frombits(binary.LittleEndian.Uint32(t.Data[i*4:]))
		}
		return out, nil
	case "F16":
		out := make([]float32, len(t.Data)/2)
		for i := range out {
			h := binary.LittleEndian.Uint16(t.Data[i*2:])
			out[i] = HalfToFloat32(h)
		}
		return out, nil
	}
	return nil, fmt.Errorf("%s: expected F32/F16, got %s", name, t.Dtype)
}

// I64 returns a tensor's data as []int64 (I64 only).
func (f *File) I64(name string) ([]int64, error) {
	t, ok := f.Tensors[name]
	if !ok {
		return nil, fmt.Errorf("no tensor %q", name)
	}
	if t.Dtype != "I64" {
		return nil, fmt.Errorf("%s: expected I64, got %s", name, t.Dtype)
	}
	out := make([]int64, len(t.Data)/8)
	for i := range out {
		out[i] = int64(binary.LittleEndian.Uint64(t.Data[i*8:]))
	}
	return out, nil
}

// HalfToFloat32 upcasts an IEEE half-precision value exactly.
func HalfToFloat32(h uint16) float32 {
	sign := float32(1)
	if h&0x8000 != 0 {
		sign = -1
	}
	exp := (h >> 10) & 0x1f
	frac := h & 0x3ff
	switch {
	case exp == 0:
		return sign * float32(frac) * float32(math.Pow(2, -24))
	case exp == 31:
		if frac == 0 {
			return sign * float32(math.Inf(1))
		}
		return float32(math.NaN())
	}
	return sign * (1 + float32(frac)/1024) * float32(math.Pow(2, float64(exp)-15))
}
